contrib/drivers/hpe/nimble/nimble.go

```diff
@@ -0,0 +1,346 @@
+		Description:      opt.GetDescription(),
+		AvailabilityZone: opt.GetAvailabilityZone(),
+	}, nil
+}
+
+func (d *Driver) CreateSnapshot(opt *pb.CreateVolumeSnapshotOpts) (*model.VolumeSnapshotSpec, error) {
+	log.Infof("%v: trying create snapshot...", DriverName)
+	poolId := opt.GetMetadata()["PoolId"]
+	snap, err := d.client.CreateSnapshot(poolId, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("%v: create snapshot success.", DriverName)
+	return &model.VolumeSnapshotSpec{
+		BaseModel: &model.BaseModel{
+			Id: opt.GetId(),
+		},
+		Name:        opt.GetName(),
+		Description: opt.GetDescription(),
+		VolumeId:    opt.GetVolumeId(),
+		Size:        0,
+		Metadata: map[string]string{
+			"SnapId": snap.Id,
+		},
+	}, nil
+}
+
+func (d *Driver) DeleteSnapshot(opt *pb.DeleteVolumeSnapshotOpts) error {
+	log.Infof("%v: trying delete snapshot...", DriverName)
+	poolId := opt.GetMetadata()["PoolId"]
+	err := d.client.DeleteSnapshot(poolId, opt)
+	if err != nil {
+		log.Errorf("%v: delete volume snapshot failed, volume snapshot id = %s , error: %v", DriverName, opt.GetId(), err)
+		return err
+	}
+	log.Infof("%v: remove volume snapshot success, volume snapshot id=%v", DriverName, opt.GetId())
+	return nil
+}
+
+func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
+	log.Infof("%v: listPools ...", DriverName)
+	var pols []*model.StoragePoolSpec
+	sp, err := d.client.ListStoragePools()
+	if err != nil {
+		return nil, err
+	}
+
+	c := d.conf
+
+	for _, pool := range sp {
+		for grpName, _ := range c.Pool {
+			if grpName == pool.Name && c.Pool[grpName].AvailabilityZone == pool.ArrayList[0].ArrayName {
+
+				pol := &model.StoragePoolSpec{
+					BaseModel: &model.BaseModel{
+						//				Id: pool.Id,
+						Id: uuid.NewV5(uuid.NamespaceOID, pool.Id).String(),
+					},
+					Name:             pool.Name,
+					Description:      pool.Description,
+					TotalCapacity:    pool.TotalCapacity / int64(math.Pow(1024, 3)), //Format to GiB
+					FreeCapacity:     pool.FreeCapacity / int64(math.Pow(1024, 3)),  //Format to GiB
+					StorageType:      "block",
+					AvailabilityZone: pool.ArrayList[0].ArrayName + "/" + pool.Name,
+					Extras:           c.Pool[grpName].Extras,
+				}
+				pols = append(pols, pol)
+				break
+			}
+		}
+	}
+
+	// Error if there is NO valid storage grp
+	if len(pols) == 0 {
+		return nil, fmt.Errorf("%v: there are no valid storage pool. Pls check driver config.\n", DriverName)
+	}
+
+	return pols, nil
+}
+
+func (d *Driver) InitializeConnection(opt *pb.CreateVolumeAttachmentOpts) (*model.ConnectionInfo, error) {
+
+	if opt.GetAccessProtocol() == ISCSIProtocol || opt.GetAccessProtocol() == FCProtocol {
+		log.Infof("%v: trying initialize connection...", DriverName)
+
+		poolId := opt.GetMetadata()["PoolId"]
+		initiatorName := opt.GetHostInfo().Initiator
+		if initiatorName == "" || opt.GetHostInfo().Ip == "" {
+			return nil, fmt.Errorf("%v: pls set initiator IQN or WWPN and IP address.\n", DriverName)
```

Comment:
This error message may not be very apt while debugging, right? We may need to log the accessprotocol too. then it will identify if its IQN or WWPN

Reply:
right, It was not kind. I changed these messages with protocol.